dotCMS/src/main/java/com/dotmarketing/util/ImportUtil.java

```diff
@@ -353,210 +329,438 @@ public static HashMap<String, List<String>> importFile(Long importId, String cur
+                if (!headers.isEmpty()) {
+
                     if (!preview) {
                         HibernateUtil.startTransaction();
                     }
+
                     String[] csvLine;
                     while (csvreader.readRecord()) {
-                        if(ImportAuditUtil.cancelledImports.containsKey(importId)){
-                            break;
-                        }
+
                         lineNumber++;
                         csvLine = csvreader.getValues();
                         LineImportResultBuilder resultBuilder = null;
 
+                        // Check for cancellation
+                        if (ImportAuditUtil.cancelledImports.containsKey(importId)) {
+                            messages.add(ValidationMessage.builder()
+                                    .type(ValidationMessageType.INFO)
+                                    .lineNumber(lineNumber)
+                                    .message("Import cancelled by user")
+                                    .build());
+                            break;
+                        }
+
                         try {
-                            lines++;
-                            Logger.debug(ImportUtil.class, "Line " + lines + ": (" + csvreader.getRawRecord() + ").");
-
-                            //Importing a line
-                            Long languageToImport = language;
-                            if ( language == -1 ) {
-                                if ( languageCodeHeaderColumn != -1 && countryCodeHeaderColumn != -1 ) {
-                                    Language dotCMSLanguage = langAPI.getLanguage( csvLine[languageCodeHeaderColumn], csvLine[countryCodeHeaderColumn] );
-                                    languageToImport = dotCMSLanguage.getId();
-                                }
-                            }
 
-                            if ( languageToImport != -1 ) {
-                                /*
-                                Verifies if there was already imported a record with the same keys.
-                                Useful to know if we have batch uploads with the same keys, mostly visible for batch content uploads with multiple languages
-                                 */
-                                boolean sameKeyBatchInsert = true;
-                                if ( keyFields != null && !keyFields.isEmpty() ) {
-                                    for ( Integer column : keyFields.keySet() ) {
-                                        Field keyField = keyFields.get( column );
-										if (!counters.matchKey(keyField.getVelocityVarName(), csvLine[column])) {
-                                            sameKeyBatchInsert = false;
-                                            break;
-                                        }
-                                    }
-                                }
+                            Logger.debug(ImportUtil.class,
```

Comment:
The methods in the class are so big

Reply:
This is related with how the `ImportUtil` already works and it is outside the scope of the ticket.